Source/WebCore/accessibility/AccessibilityObject.cpp

```diff
@@ -845,56 +845,72 @@ std::optional<BoundaryPoint> AccessibilityObject::lastBoundaryPointContainedInRe
+static IntPoint textStartPoint(const IntRect& rect, bool isFlippedWritingMode)
+{
+    if (!isFlippedWritingMode)
+        return rect.minXMinYCorner();
+    return rect.maxXMinYCorner();
+}
+
+static IntPoint textEndPoint(const IntRect& rect, bool isFlippedWritingMode)
+{
+    if (!isFlippedWritingMode)
+        return rect.maxXMaxYCorner();
+    return rect.minXMaxYCorner();
+}
+
+bool AccessibilityObject::boundaryPointsContainedInRect(const BoundaryPoint& startBoundary, const BoundaryPoint& endBoundary, const FloatRect& rect, bool isFlippedWritingMode) const
 {
     auto elementRect = boundsForRange({ startBoundary, endBoundary });
-    return rect.contains(elementRect.location() + elementRect.size());
+    return rect.contains(textEndPoint(elementRect, isFlippedWritingMode));
 }
 
 std::optional<SimpleRange> AccessibilityObject::visibleCharacterRange() const
 {
     auto range = simpleRange();
+    if (!range)
+        return std::nullopt;
+
     auto contentRect = unobscuredContentRect();
     auto elementRect = snappedIntRect(this->elementRect());
-    auto inputs = std::make_tuple(range, contentRect, elementRect);
-    if (m_cachedVisibleCharacterRangeInputs && *m_cachedVisibleCharacterRangeInputs == inputs)
-        return m_cachedVisibleCharacterRange;
-
-    auto computedRange = visibleCharacterRangeInternal(range, contentRect, elementRect);
-    m_cachedVisibleCharacterRangeInputs = inputs;
-    m_cachedVisibleCharacterRange = computedRange;
-    return computedRange;
+    return visibleCharacterRangeInternal(*range, contentRect, elementRect);
 }
 
-std::optional<SimpleRange> AccessibilityObject::visibleCharacterRangeInternal(const std::optional<SimpleRange>& range, const FloatRect& contentRect, const IntRect& startingElementRect) const
+std::optional<SimpleRange> AccessibilityObject::visibleCharacterRangeInternal(SimpleRange& range, const FloatRect& contentRect, const IntRect& startingElementRect) const
 {
-    if (!range || !contentRect.intersects(startingElementRect))
+    if (!contentRect.intersects(startingElementRect))
         return std::nullopt;
 
     auto elementRect = startingElementRect;
-    auto startBoundary = range->start;
-    auto endBoundary = range->end;
+    auto startBoundary = range.start;
+    auto endBoundary = range.end;
+
+    const auto* style = this->style();
+    bool isFlipped = style && style->writingMode().isBlockFlipped();
+    // In vertical-rl writing-modes (e.g. some Japanese text), text lays out vertically from right-to-left, meaning the the start of the text
+    // has a larger `x`-coordinate than the end.
+    bool laysOutIntoNegativeX = isFlipped && style->writingMode().isVertical();
 
     // Origin isn't contained in visible rect, start moving forward by line.
-    while (!contentRect.contains(elementRect.location())) {
+    while (!contentRect.contains(textStartPoint(elementRect, isFlipped))) {
```

Comment:
is isFlipped the right word or isVerticalWriting?

Reply:
`isFlipped` is accurate. That encapsulates `vertical-rl` and `horizontal-tb` (horizontal top-to-bottom), which both flip the text start and end points (the start is at x + width or y + height rather than at x or y), which is what we care about. We would not want to apply this logic to `vertical-lr`, which is not flipped (doing so would cause the wrong behavior).

https://github.com/WebKit/WebKit/blob/1593e890b56108f2b3ff0cb7b15739ea0db8e624/Source/WebCore/platform/text/WritingMode.h#L136